include/pybind11/pybind11.h

```diff
@@ -203,6 +199,8 @@ class cpp_function : public function {
                 a.descr = strdup(a.value.attr("__repr__")().cast<std::string>().c_str());
         }
 
+        rec->is_constructor = !strcmp(rec->name, "__init__") || !strcmp(rec->name, "__setstate__");
```

Comment:
The constructor signature rewriting got reintroduced with removal of the proxy class, so I kept this moved.